pkg/controllers/node/expiration.go

```diff
@@ -0,0 +1,62 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package node
+
+import (
+	"context"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/utils/clock"
+	"knative.dev/pkg/logging"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/samber/lo"
+
+	"github.com/aws/karpenter-core/pkg/apis/v1alpha5"
+	utilsnode "github.com/aws/karpenter-core/pkg/utils/node"
+)
+
+// Expiration is a node sub-controller that annotates or de-annotates an expired node based on TTLSecondsUntilExpired
+type Expiration struct {
```

Comment:
Should this controller live in deprovisioning? Would be great if we could figure out a good pattern to build controllers that impact deprovisioning, even if they're out of the critical path.

Reply:
This follows a similar separation as the deprovisioning controller and the termination controller. Make the controller do one thing the best it can. 

The deprovisioning controller is in charge of figuring out which nodes can be deprovisioned, and lets two other controllers (provisioning, termination) complete their reconcile flow before finishing the deprovisioning reconcile loop. 

I think this is a proper separation here as the node controllers are responsible for responding to signals within the cluster to assert a state of the node. This could include actions that don't involve deprovisioning as well. Thoughts on why you want to merge these?